### PR TITLE
cxx-qt-gen: allow closures in signal connections

### PR DESCRIPTION
**Pull request by ahayzen-kdab**
Related to #595 

Requires #690

### CODE REVIEW

File: crates/cxx-qt-gen/test_inputs/signals.rs [hunk 1]
```diff
@@ -6,20 +6,33 @@ mod ffi {
...
+
     unsafe extern "RustQt" {
         #[qobject]
         type MyObject = super::MyObjectRust;
 
         #[qsignal]
         fn ready(self: Pin<&mut MyObject>);
 
+        #[qsignal]
+        fn connected(self: Pin<&mut MyObject>, url: &QUrl);
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/test_inputs/signals.rs:27]: I think this test can go, the `data_changed` signal already includes a by-reference parameter.

Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/test_inputs/signals.rs:27]: remove this

File: crates/cxx-qt-gen/src/generator/cpp/signal.rs [hunk 2]
```diff
@@ -18,53 +22,61 @@ use syn::Result;
...
+        named_types: parameter_named_types,
+        named_types_with_self: parameter_named_types_with_self.join(", "),
+        types_with_self: parameter_types_with_self.join(", "),
+        values_with_self: parameter_values_with_self.join(", "),
+        values_with_self_as_this: parameter_values_with_self_as_this.join(", "),
     })
 }
 
 /// Generate C++ blocks for a free signal on an existing QObject (not generated by CXX-Qt), eg QPushButton::clicked
 pub fn generate_cpp_free_signal(
```
Comment (ahayzen-kdab) [inline crates/cxx-qt-gen/src/generator/cpp/signal.rs:74]: reuse free generation for normal signal generation, eg make everything go via free functions

File: crates/cxx-qt-gen/src/generator/rust/signals.rs [hunk 3]
```diff
@@ -186,6 +272,30 @@ pub fn generate_rust_signals(
                 parameter
             })
             .collect();
+        let parameters_name: Vec<Ident> = parameters_cxx
```
Comment (LeonMatthesKDAB) [inline crates/cxx-qt-gen/src/generator/rust/signals.rs:275]: See comment on parameters_name above.

**Review by LeonMatthesKDAB — CHANGES_REQUESTED** [review]
Overall this change is looking pretty good.
It's just a lot of added code, though most of it is simply generation of what we were testing in the playground.

The important thing I'd like to see tested is whether we can reduce code duplication by only using "free" connection methods.
As it seems like a heavy burden to support two very complicated generators if the free one could be used just as well for extern "RustQt". Even if it polutes the C++ namespace a bit more.

**Comment by ahayzen-kdab** [comment]
Changing this to use "free" signal generation all the time is likely the route.

File: crates/cxx-qt-gen/src/generator/cpp/property/mod.rs [hunk 4]
```diff
@@ -186,20 +186,20 @@ mod tests {
         };
         assert_str_eq!(
             header,
-            "::QMetaObject::Connection trivialPropertyChangedConnect(::rust::Fn<void(MyObject&)> func, ::Qt::ConnectionType type);"
+            "::QMetaObject::Connection trivialPropertyChangedConnect(::MyObjectCxxQtSignalHandlertrivialPropertyChanged closure, ::Qt::ConnectionType type);"
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/cpp/property/mod.rs]: It is now :-)

File: crates/cxx-qt-gen/src/generator/rust/signals.rs [hunk 5]
```diff
@@ -125,13 +175,47 @@ pub fn generate_rust_free_signal(
...
+                ) {
+                    handler.closure()(self_, #(#parameters_name),*);
+                }
+            },
+            quote! {
+                cxx_qt::assert_eq_align!(cxx_qt::CxxQtSignalHandler<#closure_struct>, usize);
+            },
+            quote! {
+                cxx_qt::assert_eq_size!(cxx_qt::CxxQtSignalHandler<#closure_struct>, [usize; 2]);
+            },
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs]: Problem here is that we go Vec<TokenStream> -> Vec<Item> and then multiple statements can't be converted to a single Item.

File: crates/cxx-qt-gen/src/generator/rust/signals.rs [hunk 6]
```diff
@@ -125,13 +175,47 @@ pub fn generate_rust_free_signal(
...
+                    type FnType = dyn FnMut(#self_type_qualified, #(#parameters_qualified_type),*);
+                }
+            },
+            quote! {
+                use core::mem::drop as #signal_handler_drop;
+            },
+            quote! {
+                fn #signal_handler_call(
+                    handler: &mut cxx_qt::CxxQtSignalHandler<#closure_struct>,
+                    self_: #self_type_qualified,
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs:206]: I've been using `self_` and `self_value` it appears, i'll see if `this` works in these places. Assume it doesn't confused C++ in the CXX generation...

[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs:206]: Hmm

```
[build]   cargo:warning=/var/home/andrew/Projects/cxx-qt/build/./cargo/build/x86_64-unknown-linux-gnu/debug/build/qml-minimal-no-cmake-c7f8f60f95a0d7a4/out/cxx-qt-gen/src/qobject.cxx.cpp:455:93: error: invalid use of 'this' in non-member function
[build]   cargo:warning=  455 |   rust$cxxqtgen1$externcxxqt$cxxbridge1$call_MyObject_signal_handler_stringChanged(handler, this);
[build]   cargo:warning=      |                                                                                             ^~~~
```

File: crates/cxx-qt-gen/src/generator/rust/signals.rs [hunk 7]
```diff
@@ -66,6 +66,30 @@ pub fn generate_rust_free_signal(
...
+            FnArg::Typed(pat_type) => {
+                if let Pat::Ident(pat_ident) = *pat_type.pat {
+                    pat_ident.ident
+                } else {
+                    unreachable!();
+                }
+            }
+            _ => unreachable!(),
+        })
+        .collect();
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs:80]: done

File: crates/cxx-qt-gen/src/generator/rust/signals.rs [hunk 8]
```diff
@@ -66,6 +66,30 @@ pub fn generate_rust_free_signal(
...
+    let parameters_qualified_type: Vec<Type> = parameters_cxx
+        .iter()
+        .cloned()
+        .map(|parameter| match parameter {
+            FnArg::Typed(pat_type) => {
+                syn_type_cxx_bridge_to_qualified(&pat_type.ty, &cxx_mappings.qualified)
+            }
+            FnArg::Receiver(receiver) => *receiver.ty,
```
[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs:88]: unreachable has now gone :-)

[the remark this answers is not in the document]
Comment (ahayzen-kdab) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs:88]: But yes the receiver should never be there so lets add an unreachable here

File: crates/cxx-qt-gen/src/generator/rust/signals.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs]: Ah, right... the Item conversion...
nevermind then...

File: crates/cxx-qt-gen/src/generator/rust/signals.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (LeonMatthesKDAB) [reply crates/cxx-qt-gen/src/generator/rust/signals.rs]: Oh, well, then we'll have to keep using `self_` or `self_value`...

**Review by LeonMatthesKDAB — COMMENTED** [review]
Honestly, Looks good to me.

Whilst the amount of code necessary for generation unfortunately has increased massively, our API footprint is actually really small.

So as long as we continue to support closures, we can try to reduce the amount of generation needed again.
Maybe we can somehow reinterpret_cast some more stuff so we don't have to generate as much boilerplate...
Though then the compiler could probably do a lot less checks which may leave us with a lot more potential for UB.

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

We can merge now and refactor/reduce code needed later

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

The new API is sane and we can always change the internals :)

**Review by LeonMatthesKDAB — APPROVED** [review]
LGTM

The new API is sane and we can always improve the internals later.

**Pull request merged**